### PR TITLE
Improve Equals(object) in SharpDx.Mathematics.

### PR DESCRIPTION
**Pull request by andrewst**
Hello,
I propose to improve the Equals in SharpDx.Mathematics

### CODE REVIEW

File: Source/SharpDX.Mathematics/BoundingFrustum.cs [hunk 1]
```diff
@@ -150,9 +150,10 @@ public bool Equals(BoundingFrustum other)
         /// </returns>
         public override bool Equals(object obj)
         {
-            if (obj != null && obj is BoundingFrustum)
-                return Equals((BoundingFrustum)obj);
-            return false;
+            if(!(obj is BoundingBox))
```
Comment (xoofx) [inline Source/SharpDX.Mathematics/BoundingFrustum.cs]: Be careful, this should be `BoudingFrustum` instead of `BoudingBox`

**Comment by andrewst** [comment]
What do you think, if add overloads (public bool Equals(ref %TYPE% other)) in all valueType?

**Comment by xoofx** [comment]
@andrewst why not... the only thing I'm worried is performance of standard Equals(%TYPE%) if it starts to use the ref version... not sure the inline will work in all cases without aggressive inlining...

**Comment by andrewst** [comment]
I testing previous version (without my change) in .net 4.5.2 x64.
Method Bool4.Equals(Bool4 other) not inlining
C# code:
```csharp
private static void TestBool4()
{
    var bool1 = new Bool4(true, false, false, false);
    var bool2 = new Bool4(true, true, false, false);
    var result = bool1.Equals(bool2);
    Console.WriteLine(result);
}
```
JIT generating:
```asm
    var bool1 = new Bool4(true, false, false, false);
00007FFE4DC85D62  sub         esp,60h  
00007FFE4DC85D65  lea         rdi,[rsp+40h]  
00007FFE4DC85D6A  mov         ecx,8  
00007FFE4DC85D6F  xor         eax,eax  
00007FFE4DC85D71  rep stos    dword ptr [rdi]  
00007FFE4DC85D73  xor         ecx,ecx  
00007FFE4DC85D75  mov         dword ptr [rsp+20h],ecx  
00007FFE4DC85D79  lea         rcx,[rsp+50h]  
00007FFE4DC85D7E  mov         edx,1  
00007FFE4DC85D83  xor         r8d,r8d  
00007FFE4DC85D86  xor         r9d,r9d  
00007FFE4DC85D89  call        00007FFE4DC854D0  <===Call first ctor
    var bool2 = new Bool4(true, true, false, false);
00007FFE4DC85D8E  xor         ecx,ecx  
00007FFE4DC85D90  mov         dword ptr [rsp+20h],ecx  
00007FFE4DC85D94  lea         rcx,[rsp+40h]  
00007FFE4DC85D99  mov         edx,1  
00007FFE4DC85D9E  mov         r8d,1  
00007FFE4DC85DA4  xor         r9d,r9d  
00007FFE4DC85DA7  call        00007FFE4DC854D0   <===Call second ctor
00007FFE4DC85DAC  lea         rcx,[rsp+50h]  
00007FFE4DC85DB1  movdqu      xmm0,xmmword ptr [rsp+40h]  
00007FFE4DC85DB7  movdqu      xmmword ptr [rsp+30h],xmm0  
00007FFE4DC85DBD  lea         rdx,[rsp+30h]  
00007FFE4DC85DC2  call        00007FFE4DC85548    <===Call Equals
00007FFE4DC85DC7  movzx       ecx,al  
00007FFE4DC85DCA  call        00007FFEAC7EA590   <===Call Console.WriteLine
00007FFE4DC85DCF  nop  
00007FFE4DC85DD0  add         rsp,60h  
00007FFE4DC85DD4  pop         rdi  
00007FFE4DC85DD5  ret   
```
if add [MethodImpl(MethodImplOptions.AggressiveInlining)] in Equals
JIT generating:
```asm
var bool1 = new Bool4(true, false, false, false);
00007FFE4DC65D62  sub         esp,50h  
00007FFE4DC65D65  lea         rdi,[rsp+30h]  
00007FFE4DC65D6A  mov         ecx,8  
00007FFE4DC65D6F  xor         eax,eax  
00007FFE4DC65D71  rep stos    dword ptr [rdi]  
00007FFE4DC65D73  xor         ecx,ecx  
00007FFE4DC65D75  mov         dword ptr [rsp+20h],ecx  
00007FFE4DC65D79  lea         rcx,[rsp+40h]  
00007FFE4DC65D7E  mov         edx,1  
00007FFE4DC65D83  xor         r8d,r8d  
00007FFE4DC65D86  xor         r9d,r9d  
00007FFE4DC65D89  call        00007FFE4DC654D0   <===Call first ctor
    var bool2 = new Bool4(true, true, false, false);
00007FFE4DC65D8E  xor         ecx,ecx  
00007FFE4DC65D90  mov         dword ptr [rsp+20h],ecx  
00007FFE4DC65D94  lea         rcx,[rsp+30h]  
00007FFE4DC65D99  mov         edx,1  
00007FFE4DC65D9E  mov         r8d,1  
00007FFE4DC65DA4  xor         r9d,r9d  
00007FFE4DC65DA7  call        00007FFE4DC654D0   <===Call second ctor
    var result = bool1.Equals(bool2);
00007FFE4DC65DAC  lea         rcx,[rsp+30h]  
00007FFE4DC65DB1  mov         eax,dword ptr [rcx]  
00007FFE4DC65DB3  mov         edx,dword ptr [rcx+4]  
00007FFE4DC65DB6  mov         r8d,dword ptr [rcx+8]  
00007FFE4DC65DBA  mov         ecx,dword ptr [rcx+0Ch]  
00007FFE4DC65DBD  test        eax,eax  
00007FFE4DC65DBF  seta        al  
00007FFE4DC65DC2  movzx       eax,al  
00007FFE4DC65DC5  mov         r9d,dword ptr [rsp+40h]  
00007FFE4DC65DCA  test        r9d,r9d  
00007FFE4DC65DCD  seta        r9b  
00007FFE4DC65DD1  movzx       r9d,r9b  
00007FFE4DC65DD5  cmp         eax,r9d  
00007FFE4DC65DD8  jne         00007FFE4DC65E29  
00007FFE4DC65DDA  test        edx,edx  
00007FFE4DC65DDC  seta        al  
00007FFE4DC65DDF  movzx       eax,al  
00007FFE4DC65DE2  mov         edx,dword ptr [rsp+44h]  
00007FFE4DC65DE6  test        edx,edx  
00007FFE4DC65DE8  seta        dl  
00007FFE4DC65DEB  movzx       edx,dl  
00007FFE4DC65DEE  cmp         eax,edx  
00007FFE4DC65DF0  jne         00007FFE4DC65E29  
00007FFE4DC65DF2  test        r8d,r8d  
00007FFE4DC65DF5  seta        al  
00007FFE4DC65DF8  movzx       eax,al  
00007FFE4DC65DFB  mov         edx,dword ptr [rsp+48h]  
00007FFE4DC65DFF  test        edx,edx  
00007FFE4DC65E01  seta        dl  
00007FFE4DC65E04  movzx       edx,dl  
00007FFE4DC65E07  cmp         eax,edx  
00007FFE4DC65E09  jne         00007FFE4DC65E29  
00007FFE4DC65E0B  test        ecx,ecx  
00007FFE4DC65E0D  seta        cl  
00007FFE4DC65E10  movzx       ecx,cl  
00007FFE4DC65E13  mov         eax,dword ptr [rsp+4Ch]  
00007FFE4DC65E17  test        eax,eax  
00007FFE4DC65E19  seta        al  
00007FFE4DC65E1C  movzx       eax,al  
00007FFE4DC65E1F  cmp         ecx,eax  
00007FFE4DC65E21  sete        cl  
00007FFE4DC65E24  movzx       ecx,cl  
00007FFE4DC65E27  jmp         00007FFE4DC65E2B  
00007FFE4DC65E29  xor         ecx,ecx  
00007FFE4DC65E2B  call        00007FFEAC7EA590   <===Call Console.WriteLine
00007FFE4DC65E30  nop  
00007FFE4DC65E31  add         rsp,50h  
00007FFE4DC65E35  pop         rdi  
00007FFE4DC65E36  ret  
```

I think adding AggressiveInlining attribute in all overloads Equals methods
Example:
```csharp
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool Equals(ref Bool4 other)
        {
            return other.X == X && other.Y == Y && other.Z == Z && other.W == W;
        }
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool Equals(Bool4 other)
        {
            return Equals(ref other);
        }
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public override bool Equals(object value)
        {
            if (!(value is Bool4))
                return false;

            var strongValue = (Bool4)value;
            return Equals(ref strongValue);
        }
```

**Comment by xoofx** [comment]
Thanks, this is merged.

**Pull request closed without merging**